ecosystem/platform/server2/app/models/user.rb

```diff
@@ -0,0 +1,137 @@
+
+# Copyright (c) Aptos
+# SPDX-License-Identifier: Apache-2.0
+
+class User < ApplicationRecord
+  include RailsStateMachine::Model
+
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :recoverable,
+  devise :database_authenticatable,
+         :rememberable, :trackable, :validatable,
+         :omniauthable, omniauth_providers: %i[discord github],
+                        authentication_keys: [:username]
+
+  validates :username, uniqueness: { case_sensitive: false }, allow_nil: true
+  validates :email, uniqueness: { case_sensitive: false }, allow_nil: true
+
+  validate_hex :mainnet_address
+
+  has_many :authorizations, dependent: :destroy
+
+  # https://github.com/makandra/rails_state_machine
+  # TODO: this state machine
+  state_machine :kyc_status do
+    state :not_started, initial: true
+  end
+
+  def self.from_omniauth(auth, current_user = nil)
+    # find an existing user or create a user and authorizations
+    # schema of auth https://github.com/omniauth/omniauth/wiki/Auth-Hash-Schema
+
+    # returning users
+    authorization = Authorization.find_by(provider: auth.provider, uid: auth.uid)
+    return authorization.user if authorization
+
+    # if user is already logged in, add new oauth to existing user
+    if current_user
+      current_user.add_oauth_authorization(auth).save!
+      return current_user
+    end
+
+    # Totally new user
+    username = auth['info'].fetch('nickname') || auth['info'].fetch('name')
+    create_new_user_from_oauth(auth, username, auth['info']['email'])
+  end
+
+  def self.create_new_user_from_oauth(auth, username, email)
+    # If this username is taken, add random suffix to ensure account is created
+    # TODO: while unlikely, this does expose an interesting avenue for DOS.
+    username = "#{username}-#{rand(10**12)}" while User.where(username:).exists?
+
+    # If this email is taken, dont save it
+    email = nil if User.where(email:).exists?
+
+    user = User.new({
+                      username:,
+                      email:,
+                      password: Devise.friendly_token(32)
+                    })
+    # user.skip_confirmation! if %w[google].include?(auth.provider)
+    user.add_oauth_authorization(auth)
+    user.save
+    user
+  end
+
+  # Maintaining state if a user was not able to be saved
+  # def self.new_with_session(params, session)
+  #   super.tap do |user|
+  #     if (data = session['devise.oauth.data'])
+  #       user.email = data['info']['email'] if user.email.blank?
+  #       user.add_oauth_authorization(data)
+  #     end
+  #   end
+  # end
+
+  def providers
+    authorizations.map(&:provider)
+  end
+
+  def add_oauth_authorization(data)
+    expires_at = begin
+      Time.at(data['credentials']['expires_at'])
+    rescue StandardError
+      nil
+    end
+    # byebug
+    auth = {
+      provider: data['provider'],
+      uid: data['uid'],
+      token: data['credentials']['token'],
```

Comment:
Will it crash if the data doesn't match this structure?

Reply:
yes, but this structure is taken from their API spec